### PR TITLE
Fix a bug handling Jinja "{% set %}" blocks with a templated block inside

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made

Makes progress on #2235 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by barrywhart** [comment]
In the current state of this PR, `sqlfluff parse` on this file throws an exception:
```
{% set my_query %}
  select 1 from foobarfoobarfoobarfoobar_{{ "dev" }}
{% endset %}
{{ my_query }}
```

File: test/core/templaters/jinja_test.py
Comment (barrywhart) [inline test/core/templaters/jinja_test.py:512]: Note the `{{ }}` block on this line inside the `{% set %}` block. This was previously not handled correctly and caused a runtime error.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:73]: Previously, each output entry was terminated by a `\0` character. In order to handle the new case, it proved better to **prefix** each output entry with a `\0` instead, hence I refactored the loop.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:173]: This function replaces uses of `uuid.uuid4().hex`. There was really no need to use UUIDs, and changing to a deterministic counter makes the tests easier to understand and debug.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:195]: Here's where I moved the `\0` from the end of each chunk to the beginning.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:142]: I tweaked this function to optionally take a `slice_idx` so I could reuse it to process output from a `{% set %}` trace. Previously it always used `self.program_counter`.

File: test/core/templaters/jinja_test.py
Comment (tunetheweb) [reply test/core/templaters/jinja_test.py:512]: Might be good to link to the docs on this, as I didn't realise this was valid syntax, but looks like it is: https://jinja.palletsprojects.com/en/3.0.x/templates/#block-assignments

Comment (tunetheweb) [reply test/core/templaters/jinja_test.py:512]: Ah ignore me. See you added that comment in the code itself on line 314.

Comment (barrywhart) [reply test/core/templaters/jinja_test.py:514]: Done!

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/slicers/tracer.py:81]: This only covers the `{% set... %}` case but looking at https://jinja.palletsprojects.com/en/3.0.x/templates/ there are lots of similar cases (`for`, `call`, `macro`). Should we include them in this?

Or at least add a comment saying they are not supported for now?

Or where they already covered (I'd be really surprised if `for` was not covered already)?

File: test/core/templaters/jinja_test.py
Comment (barrywhart) [reply test/core/templaters/jinja_test.py:514]: No worries, it's often a good idea to document both the tests and the code, to help connect the dots.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [reply src/sqlfluff/core/templaters/slicers/tracer.py:81]: `{% for %}` is covered by the general block handling -- see the `block_types` lookup elsewhere in the file. Jinja is helping us with some of this by telling us the element type. TBH, I don't know if `call` and `macro` work. SInce I don't know for sure, I would rather not add comments about those.

Comment (barrywhart) [reply src/sqlfluff/core/templaters/slicers/tracer.py:81]: Note that `tracer.py` is the primary (only) component used in the past few releases for Jinja and dbt template slicing, so it's fairly robust. Given the extreme amount of templating used in dbt projects, I think it's pretty robust, just can't say for sure. AFAIK, this issue was the first one raised since we switched over in late October/early November.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/slicers/tracer.py:301]: Can we update this comment since not using UUID's anymore?

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/slicers/tracer.py:97]: Just above this are two more UUID references in comments (I think). Should they be updated.

Sorry can't comment on unchanged lines.

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (tunetheweb) [inline src/sqlfluff/core/templaters/slicers/tracer.py:109]: Just below this is another UUID reference.

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Can't say I honestly "get" all the code but happy with the test cases.

There's a few UUID references to clean up but after that should be good to merge.

**Comment by barrywhart** [comment]
Ok, I believe I've updated/removed all the references to UUIDs. Thanks for the review!

**Review by tunetheweb — APPROVED** [review]

File: src/sqlfluff/core/templaters/slicers/tracer.py
Comment (barrywhart) [reply src/sqlfluff/core/templaters/slicers/tracer.py:73]: Hmm, I tried this, and it wasn't working. Perhaps the objects in the list are different types? I think `finditer` returns match objects, while `findall` returns tuples of strings.

```
        trace_entries = regex.findall(r"\0", trace_template_output)
        for match_idx, match in enumerate(trace_entries):
>           pos1 = match.span()[0]
E           AttributeError: 'str' object has no attribute 'span'
```

Comment (jpy-git) [reply src/sqlfluff/core/templaters/slicers/tracer.py:73]: ahh sorry you're right findall returns `list[AnyStr] | list[tuple[AnyStr, ...]]` so ignore me 😄

**Pull request merged**
